src/gsHSplines/gsTHBSplineBasis.hpp

```diff
@@ -1057,6 +1057,81 @@ void gsTHBSplineBasis<d,T>::globalRefinement(const gsMatrix<T> & thbCoefs,
 }
 
 
+template<short_t d, class T>
+void gsTHBSplineBasis<d,T>::active_into(const gsMatrix<T>& u, gsMatrix<index_t>& result) const
+{
+    gsMatrix<T> currPoint;
+    point low, upp, cur;
+    const int maxLevel = this->m_tree.getMaxInsLevel();
+
+    std::vector<std::vector<index_t> > temp_output;//collects the outputs
+    temp_output.resize( u.cols() );
+    size_t sz = 0;
+
+    for(index_t p = 0; p < u.cols(); p++) //for all input points
+    {
+        currPoint = u.col(p); 
+        for(short_t i = 0; i != d; ++i)
+            low[i] = m_bases[maxLevel]->knots(i).uFind( currPoint(i,0) ).uIndex();
+
+        // Identify the level of the point
+        const int lvl = this->m_tree.levelOf(low, maxLevel);
+
+        for(int i = 0; i <= lvl; i++)
+        {
+            m_bases[i]->active_cwise(currPoint, low, upp);
+            cur = low;
+            do
+            {
+                typename CMatrix::const_iterator it =
+                    m_xmatrix[i].find_it_or_fail( m_bases[i]->index(cur) );
+
+                if( it != m_xmatrix[i].end() )// if index is found
+                {
+                    const index_t act = this->m_xmatrix_offset[i] + (it - m_xmatrix[i].begin());
+
+                    if (this->m_is_truncated[act] == -1) 
+                    {
+                        temp_output[p].push_back(act);
+                    }
+                    else 
+                    {
+                        const gsSparseVector<T>& coefs = getCoefs(act);
+                        const gsTensorBSplineBasis<d, T>& base =
+                            *this->m_bases[this->m_is_truncated[act]];
+
+                        gsMatrix<index_t> ind;
```

Comment:
It is more efficient to declare all gsMatrix<..> objects outside all loops.
This way the memory is allocated once and reused throught the loop - it can make a significant changes in speed.

Reply:
Thanks for your advice. I've learned.